### ISSUE TITLE
Autodoc - things are messed up for inherited members when using annotations

### ISSUE THREAD

**Issue by alxroyer** [sphinx-doc/sphinx#11387]
### Describe the bug

I get strange things for inherited members with annotations:
- The `:no-inherited-members:` seems to have no effect on annotated members.
- Things get even worse when using multiple inheritance.

### How to Reproduce

3 files, in the same directory:
- 'conf.py':
  ```python
  extensions = ["sphinx.ext.autodoc"]

  # Hack `get_class_members()` usage in `sphinx.ext.autodoc` to watch out what is going on.
  import sphinx.ext.autodoc
  import sphinx.ext.autodoc.importer
  def get_class_members(subject, objpath, attrgetter, inherit_docstrings=True):
      members = sphinx.ext.autodoc.importer.get_class_members(subject, objpath, attrgetter, inherit_docstrings=inherit_docstrings)

      for name in members:
          if name in ("foo1", "foo2", "bar1", "bar2"):
              print(f"{subject.__qualname__}.{name} = {members[name]!r}, class_={members[name].class_!r}")
      if hasattr(subject, "__annotations__"):
          print(f"{subject.__qualname__}.__annotations__={subject.__annotations__!r}")
      else:
          print(f"No annotations for {subject.__qualname__}")

      return members
  sphinx.ext.autodoc.get_class_members = get_class_members
  ```
- 'index.rst':
  ```rst
  .. automodule:: sample
     :members:
     :no-inherited-members:
  ```
- 'sample.py':
```python
  """
  Class hierarchy:

  .. code-block::

      [A1] [A2]
        |\  |
        | \ |
        |  \|
      [B1] [B2]
        |   |
        |   |
        |   |
      [C1] [C2]
  """

  class A1:
      """This is A1."""
      def __init__(self):
          #: This is A1.foo1.
          self.foo1 = 0
          #: This is A1.foo2.
          self.foo2: int = 0
  class A2:
      """This is A2."""
      def __init__(self):
          #: This is A2.bar1.
          self.bar1 = 0
          #: This is A2.bar2.
          self.bar2: int = 0

  class B1(A1):
      """This is B1."""
  class B2(A1, A2):
      """This is B2."""

  class C1(B1):
      """This is C1."""
  class C2(B2):
      """This is C2."""
  ```

Build command, from the same directory:
```bash
$ python -m sphinx -b html . out/html
Running Sphinx v5.3.0
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 1 source files that are out of date
updating environment: [new config] 1 added, 0 changed, 0 removed
A1.foo1 = ('foo1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
A1.foo2 = ('foo2', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
No annotations for A1
A2.bar1 = ('bar1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A2'>
A2.bar2 = ('bar2', <object object at 0x0000029D860AA140>), class_=<class 'sample.A2'>
No annotations for A2
B1.foo2 = ('foo2', <object object at 0x0000029D860AA140>), class_=<class 'sample.B1'>
B1.foo1 = ('foo1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
B1.__annotations__={'foo2': 'int'}
B2.foo2 = ('foo2', <object object at 0x0000029D860AA140>), class_=<class 'sample.B2'>
B2.foo1 = ('foo1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
B2.bar2 = ('bar2', <object object at 0x0000029D860AA140>), class_=<class 'sample.A2'>
B2.bar1 = ('bar1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A2'>
B2.__annotations__={'foo2': 'int'}
C1.foo2 = ('foo2', <object object at 0x0000029D860AA140>), class_=<class 'sample.C1'>
C1.foo1 = ('foo1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
C1.__annotations__={'foo2': 'int'}
C2.foo2 = ('foo2', <object object at 0x0000029D860AA140>), class_=<class 'sample.C2'>
C2.bar2 = ('bar2', <object object at 0x0000029D860AA140>), class_=<class 'sample.C2'>
C2.foo1 = ('foo1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A1'>
C2.bar1 = ('bar1', <object object at 0x0000029D860AA140>), class_=<class 'sample.A2'>
C2.__annotations__={'foo2': 'int', 'bar2': 'int'}
...
```

### Environment Information

```text
Platform:              win32; (Windows-10-10.0.19041-SP0)
Python version:        3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 18:58:18) [MSC v.1900 64 bit (AMD64)])
Python implementation: CPython
Sphinx version:        5.3.0
Docutils version:      0.18.1
Jinja2 version:        3.1.2
```

### Sphinx extensions

```python
`sphinx.ext.autodoc` only (see 'conf.py' above).
```

**Comment by alxroyer** [comment]
With the given example:
- Simple inheritance:
  - `A1.foo1` is not re-documented in `B1` and `C1` as expected.
  - But the annotated version `A1.foo2` is unexpectedly re-documented in both `B1` and `C1`.
- Multiple inheritance:
  - Both `A1.foo2` and `A2.bar2` are unexpectedly re-documented in `C2`.
  - Surprisingly, only `A1.foo2` is unexpectedly re-documented in `B2`.

**Comment by alxroyer** [comment]
I bet something is going wrong regarding either `__annotations__` or `ObjectMember.class_` computations,
that's the reason why I've included a `get_class_members()` hack in my 'conf.py' in order to show a couple of things about it:
- `B2` only has annotations info for `foo2`, while `C2` has for both `foo2` and `bar2`.
- Untyped members point to `Ax` classes, while typed members point to final `Bx` or `Cx` classes. I guess all members, either typed or not, should point to their actual definition class, i.e. `Ax`.

**Comment by picnixz** [comment]
The problem is likely due to your Python version. I confirmed the issue occurs for Python 3.8 and Python 3.9 but for Python 3.10 (and Sphinx 6.2.1) it works:

```
A1.foo1 = ('foo1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>                                                 
A1.foo2 = ('foo2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
A1.__annotations__={}
A2.bar1 = ('bar1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
A2.bar2 = ('bar2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
A2.__annotations__={}
B1.foo2 = ('foo2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
B1.foo1 = ('foo1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
B1.__annotations__={}
B2.foo2 = ('foo2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
B2.foo1 = ('foo1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
B2.bar2 = ('bar2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
B2.bar1 = ('bar1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
B2.__annotations__={}
C1.foo2 = ('foo2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
C1.foo1 = ('foo1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
C1.__annotations__={}
C2.foo2 = ('foo2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
C2.foo1 = ('foo1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A1'>
C2.bar2 = ('bar2', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
C2.bar1 = ('bar1', <object object at 0x7fdf744dd4f0>), class_=<class 'testmodule.A2'>
C2.__annotations__={}
```

**Comment by alxroyer** [comment]
Good! 👍

For the memo, until one can align with a version that includes the fix, here's the workaround I use with sphinx v5.3.0:
```python
# conf.py
import sphinx.ext.autodoc
import sphinx.ext.autodoc.importer
import sphinx.pycode
import sphinx.util.inspect

extensions = ["sphinx.ext.autodoc"]


# Hack `get_class_members()` usage in `sphinx.ext.autodoc` to watch out what is going on
# and work it around.
def get_class_members(subject, objpath, attrgetter, inherit_docstrings=True):
    members = sphinx.ext.autodoc.importer.get_class_members(subject, objpath, attrgetter, inherit_docstrings=inherit_docstrings)

    # ==========
    # Workaround for [sphinx#11387](https://github.com/sphinx-doc/sphinx/issues/11387):
    for mro_class in sphinx.util.inspect.getmro(subject):  # type: type
        # Don't proceed with builtin superclasses.
        if mro_class.__module__ == "builtins":
            continue

        # Inspired from `sphinx.ext.autodoc.importer.get_class_members()`.
        analyzer = sphinx.pycode.ModuleAnalyzer.for_module(mro_class.__module__)
        analyzer.analyze()
        for (ns, name), docstring in analyzer.attr_docs.items():
            if (
                # Is the attribute name of those described in `members`?
                (name in members)
                # Is the given attribute name actually an attribute of `mro_class`?
                and (ns == mro_class.__name__)
                # Is it the same docstring? If not, the attribute may have been redefined, with a different documentation normally.
                and (members[name].docstring == "\n".join(docstring))
                # Is it really new information? If not, no need to log anything.
                and (members[name].class_ is not mro_class)
            ):
                print(f"Fixing definition class for {subject.__qualname__}.{name} from {members[name].class_!r} to {mro_class!r}")
                members[name].class_ = mro_class
    # ==========

    for name in members:
        if name in ("foo1", "foo2", "bar1", "bar2"):
            print(f"{subject.__qualname__}.{name} = {members[name]!r}, class_={members[name].class_!r}")
    if hasattr(subject, "__annotations__"):
        print(f"{subject.__qualname__}.__annotations__={subject.__annotations__!r}")
    else:
        print(f"No annotations for {subject.__qualname__}")

    return members


sphinx.ext.autodoc.get_class_members = get_class_members
```

**Comment by AA-Turner** [comment]
Fixed by #11917
